Add developer guide (#525)
HumanCompatibleAI/imitation: docs/development/developer.rst

All of the algorithms provide the ``train()`` method for training an agent and/or a reward network.

All the available algorithms are present in ``algorithms/`` with each algorithm in a distinct file. 
Adversarial algorithms like AIRL and GAIL are present in ``algorithms/adversarial``.

Data
----
.. automodule:: imitation.data
    :noindex:

    ``data.wrapper.BufferingWrapper``: Wraps a ``VecEnv`` to save the trajectories from all the environments
    in a buffer.

    ``data.wrapper.RolloutInfoWrapper``: Wraps a ``gym.Env`` to log the original observations and rewards recieved from 
    the environment. The observations and rewards of the entire episode is logged in the ``info`` dictionary with the
    key "rollout" of the step where the episode ends. This wrapper is useful to save rollout trajectories, especially
    in cases where you want to bypass the reward and/or observation overrides from other wrappers. 
    See ``data.rollout.unwrap_traj`` for details and ``scripts/train_rl.py`` for example usecase.

    ``data.rollout.rollout``: Generates rollout by taking in any policy as input along with the environment. 

> .. Envs

Reviewer: Should probably delete this rather than leave it commented out? I'm fine not describing the environments in detail as https://github.com/HumanCompatibleAI/imitation/pull/541 in the process of moving them to another repo anyway.

.. ----
.. .. automodule:: imitation.envs
..     :noindex:

> .. Policies

Reviewer: I lean towards describing the policies module if only in brief, but it is a bit of a hodge-podge of fairly unrelated modules (the only one of which that plays a critical role is `policies/serialize.py`), so skipping it is not crazy.

.. --------
.. .. automodule:: imitation.policies
..     :noindex:

Rewards
-------
.. automodule:: imitation.rewards
    :noindex:

> 

Reviewer: Worth discussing `RewardNet` in brief? That's probably the core functionality added by this module.

    ``rewards.reward_wrapper.RewardVecEnvWrapper``: This class wraps a ``VecEnv`` with a custom ``RewardFn``. 
    The default reward function of the environment is overridden with the passed reward function 
    and the original rewards are stored in the ``info_dict`` with the ``original_env_rew`` key. 
    This class is used to override the original reward function of an environment with a learned 
    reward function from the reward learning algorithms like preference comparisons.

> Scripts

Reviewer: This section is a good summary of the key features in Sacred, thanks :)

-------

> .. .. automodule:: imitation.scripts

Reviewer: delete rather than comment out?

..     :noindex:

We use Sacred to provide a command-line interface to run the experiments. The scripts to run the end-to-end experiments are
available in ``scripts/``. You can take a look at the following doc links to understand how to use Sacred:

> - `Experiment Overview <https://sacred.readthedocs.io/en/stable/experiment.html>`_: Explains how to create and run experiments. All the experiment objects are defined in ``scripts/config`` in the respective algorithm name file and their ``main`` functions are defined in algorithm name file in ``scripts/``. For example, ``train_rl_ex`` object is defined in ``scripts.config.train_rl`` and its main function is in ``scripts.train_rl``.

Reviewer: ```suggestion
- `Experiment Overview <https://sacred.readthedocs.io/en/stable/experiment.html>`_: Explains how to create and run experiments. Each script, defined in ``scripts/``, has a corresponding experiment object, defined in ``scripts/config``, with the experiment object and Python source files named after the algorithm(s) supported. For example, the ``train_rl_ex`` object is defined in ``scripts.config.train_rl`` and its main function is in ``scripts.train_rl``.
```

- `Ingredients <https://sacred.readthedocs.io/en/stable/ingredients.html>`_: Explains how to use ingredients to avoid code duplication across experiments. The ingredients used in our experiments are defined in ``scripts/common/``:
  

>   .. autosummary:: 

Reviewer: This is good! One of the most confusing parts of `imitation` for newcomers, will definitely help new devs.

    imitation.scripts.common.common
    imitation.scripts.common.demonstrations
    imitation.scripts.common.reward
    imitation.scripts.common.rl
    imitation.scripts.common.train
    imitation.scripts.common.wb    

- `Configurations <https://sacred.readthedocs.io/en/stable/configuration.html>`_: Explains how to use configurations to parametrize runs. The configurations for different algorithms are defined in their file in ``scripts/``. Some of the commonly used configs and ingredients used across algorithms are defined in ``scripts/common/``.

- `Command-Line Interface <https://sacred.readthedocs.io/en/stable/command_line.html>`_: Explains how to run the experiments through the command-line interface. Also note the section on how to `print configs <https://sacred.readthedocs.io/en/stable/command_line.html#print-config>`_ to verify the configurations used for the run.

- `Controlling Randomness <https://sacred.readthedocs.io/en/stable/randomness.html>`_: Explains how to control randomness by seeding experiments through Sacred.

> .. Util

Reviewer: I think it's worth having a `util` section. Needn't describe all of it, but `imitation.util.logger` is actually both quite important (used all over the code) and a common point of confusion for people. Could also give one-sentence descriptions of the other modules.

    the environment. The observations and rewards of the entire episode are logged in the ``info`` dictionary with the
    key ``"rollout"``, in the final time step of the episode. This wrapper is useful for saving rollout trajectories, especially
    in cases where you want to bypass the reward and/or observation overrides from other wrappers. 
    See ``data.rollout.unwrap_traj`` for details and ``scripts/train_rl.py`` for an example use case.

    ``data.rollout.rollout``: Generates rollout by taking in any policy as input along with the environment. 

Policies
--------

The ``imitation.policies`` subpackage contains the following modules:

- ``policies.base``: defines commonly used policies across the library like ``FeedForward32Policy``, ``SAC1024Policy``, ``NormalizeFeaturesExtractor``, etc.
- ``policies.exploration_wrapper``: defines the ``ExplorationWrapper`` class that wraps a policy to create a partially randomized policy useful for exploration.
- ``policies.replay_buffer_wrapper``: defines the ``ReplayBufferRewardWrapper`` to wrap a replay buffer that returns transitions with rewards specified by a reward function.

> - ``policies.serialize``: defines various functions to save and load serialized policies from the disk or the huggingface hub.

Reviewer: ```suggestion
- ``policies.serialize``: defines various functions to save and load serialized policies from the disk or the Hugging Face hub.
```

Regularization
--------------

> The ``imitation.regularization`` subpackage provides API for creating neural network regularizers. It provides classes such as 

Reviewer: ```suggestion
The ``imitation.regularization`` subpackage provides an API for creating neural network regularizers. It provides classes such as 
```

> ``regularizers.LpRegularizer`` and ``regularizers.WeightDecayRegularizer`` to regularizer the loss function and the weights of

Reviewer: ```suggestion
``regularizers.LpRegularizer`` and ``regularizers.WeightDecayRegularizer`` to regularize the loss function and the weights of
```

a network, respectively. The ``updaters.IntervalParamScaler`` class also provides support to scale the lambda hyperparameter

> of a regularizer if it lies outside an interval.

Reviewer: ```suggestion
of a regularizer up when the ratio of validation to training loss is above an upper bound, and scales it down when the ratio drops below a lower bound.
```

Rewards
-------

The ``imitation.rewards`` subpackage contains code related to building, serializing, and loading reward networks. 
Some of the classes include:

> ``rewards.reward_nets.RewardNet``: is the base reward network class. Reward networks can take state, action, and the next state as input to predict the reward. The ``forward`` method is used while training the network, whereas the ``predict`` method is used during evaluation.

Reviewer: Make it a list?

```suggestion
- ``rewards.reward_nets.RewardNet``: is the base reward network class. Reward networks can take state, action, and the next state as input to predict the reward. The ``forward`` method is used while training the network, whereas the ``predict`` method is used during evaluation.
```

> ``rewards.reward_nets.BasicRewardNet``: builds an MLP reward network.

Reviewer: ```suggestion
- ``rewards.reward_nets.BasicRewardNet``: builds a MLP reward network.
```

> ``rewards.reward_nets.CnnRewardNet``: builds a CNN based reward network.

Reviewer: ```suggestion
- ``rewards.reward_nets.CnnRewardNet``: builds a CNN based reward network.
```

> ``rewards.reward_nets.RewardEnsemble``: builds an ensemble of reward networks.

Reviewer: ```suggestion
- ``rewards.reward_nets.RewardEnsemble``: builds an ensemble of reward networks.
```

> ``rewards.reward_wrapper.RewardVecEnvWrapper``: This class wraps a ``VecEnv`` with a custom ``RewardFn``. 

Reviewer: ```suggestion
- ``rewards.reward_wrapper.RewardVecEnvWrapper``: This class wraps a ``VecEnv`` with a custom ``RewardFn``. 
```

    imitation.scripts.common.reward
    imitation.scripts.common.rl
    imitation.scripts.common.train
    imitation.scripts.common.wb

- `Configurations <https://sacred.readthedocs.io/en/stable/configuration.html>`_: Explains how to use configurations to parametrize runs. The configurations for different algorithms are defined in their file in ``scripts/``. Some of the commonly used configs and ingredients used across algorithms are defined in ``scripts/common/``.

- `Command-Line Interface <https://sacred.readthedocs.io/en/stable/command_line.html>`_: Explains how to run the experiments through the command-line interface. Also, note the section on how to `print configs <https://sacred.readthedocs.io/en/stable/command_line.html#print-config>`_ to verify the configurations used for the run.

- `Controlling Randomness <https://sacred.readthedocs.io/en/stable/randomness.html>`_: Explains how to control randomness by seeding experiments through Sacred.

Util
----

``imitation.util.logger.HierarchicalLogger``: A logger that supports contexts for accumulating the mean of values of all the logged keys. 

> The logger internally maintains two separate ``stable_baselines3.common.logger.Logger`` for logging the raw and the mean values.

Reviewer: ```suggestion
The logger internally maintains one ``stable_baselines3.common.logger.Logger`` object for logging the mean values, and one ``Logger`` object for the raw values for each context.
```

The ``accumulate_means`` context cannot be called inside an already open ``accumulate_means`` context. 

> The ``imitation.util.logger.configure`` function can be used to construct a ``HierarchicalLogger`` object easily.

Reviewer: ```suggestion
The ``imitation.util.logger.configure`` function can be used to easily construct a ``HierarchicalLogger`` object.
```
